### PR TITLE
Implement the remaining mouse functions.

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/platform/w3dengine/client/w3dmouse.cpp [hunk 1]
```diff
@@ -97,19 +101,133 @@ void W3DMouse::Init()
...
+
+            if (g_theDisplay != nullptr && g_theDisplay->Get_Windowed() != 0) {
+                POINT point;
+                GetCursorPos(&point);
+                ScreenToClient(g_applicationHWnd, &point);
+                DX8Wrapper::Get_D3D_Device8()->SetCursorPosition(point.x, point.y, D3DCURSOR_IMMEDIATE_UPDATE);
+            }
+
+            if (m_D3DCursorSurfaceCount > 1) {
+                int time = rts::Get_Time();
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dmouse.cpp]: `Get_Time` returns `unsigned int`.

File: src/platform/w3dengine/client/w3dmouse.cpp [hunk 2]
```diff
@@ -97,19 +101,133 @@ void W3DMouse::Init()
...
+                unk4 = fmod((time - m_D3DCursorLastDrawn) * unk7 + unk4, m_D3DCursorSurfaceCount);
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dmouse.cpp]: Perhaps call these `m_unk4`, `m_unk5`, `m_unk7` for now, so we can at least see in review they are class members.

File: src/platform/w3dengine/client/w3dmouse.cpp [hunk 3]
```diff
@@ -97,19 +101,133 @@ void W3DMouse::Init()
...
+                    unk5 = unk4;
+                    DX8Wrapper::Get_D3D_Device8()->SetCursorProperties(
+                        m_hotSpotX, m_hotSpotY, m_D3DCursorSurfaces[unk5]->Peek_D3D_Surface());
+                }
+            }
+        }
+    } else if (m_currentRedrawMode == RM_POLYGON) {
+        Image *cursor = s_PolyMouseAssets[m_currentPolyCursor];
+
+        if (cursor != 0) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dmouse.cpp]: nullptr

File: src/platform/w3dengine/client/w3dmouse.cpp [hunk 4]
```diff
@@ -485,17 +608,37 @@ bool W3DMouse::Load_D3D_Cursor_Textures(MouseCursor cursor)
...
-#endif
+    if (m_cursorInfo[cursor].directions > 1 && g_theInGameUI != nullptr && g_theInGameUI->Is_Scrolling()) {
+        Coord2D scroll = g_theInGameUI->Get_Scroll_Amount();
+
+        if (scroll.x == 0 && scroll.y == 0) {
+            m_cursorDirection = 0;
+        } else {
+            scroll.Normalize();
+            uint32_t directions = m_cursorInfo[m_currentCursor].directions;
+            m_cursorDirection = (fmod(GameMath::Atan2(scroll.y, scroll.x) + 6.28318530717958f, 6.28318530717958f)
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dmouse.cpp]: `PI * 2` several times

File: src/game/client/input/mouse.cpp [hunk 5]
```diff
@@ -773,3 +773,145 @@ void Mouse::Draw_Cursor_Text() const
...
+    m_isTooltipEmpty = tooltip.Is_Empty();
+    m_tooltipDelay = delay;
+    bool scale_changed = false;
+
+    if (!tooltip.Is_Empty()) {
+        if (scale != m_tooltipScale) {
+            scale_changed = true;
+            unsigned int length = (g_theDisplay->Get_Width() * m_tooltipWidth * scale);
+
+            if (length < 10) {
```
Comment (xezon) [inline src/game/client/input/mouse.cpp:788]: `< 120` ?

File: src/game/client/input/mouse.cpp [hunk 6]
```diff
@@ -773,3 +773,145 @@ void Mouse::Draw_Cursor_Text() const
...
+                m_tooltipColorTextCopy.green = GameMath::Fast_To_Int_Truncate(color->green * 255.0f);
+                m_tooltipColorTextCopy.blue = GameMath::Fast_To_Int_Truncate(color->blue * 255.0f);
+            }
+
+            m_tooltipColorTextCopy.alpha = m_tooltipColorText.alpha;
+        }
+
+        if (m_useTooltipAltBackColor) {
+            if (m_adjustTooltipAltColor) {
+                m_tooltipColorBackgroundCopy.red = GameMath::Fast_To_Int_Truncate(color->red * 255.0f * 0.5f);
```
Comment (xezon) [inline src/game/client/input/mouse.cpp:820]: `* 0.5f` here and `/ 2.0f` above?

File: src/game/client/input/mouse.cpp [hunk 7]
```diff
@@ -773,3 +773,145 @@ void Mouse::Draw_Cursor_Text() const
...
+        m_tooltipColorTextCopy = m_tooltipColorText;
+        m_tooltipColorBackgroundCopy = m_tooltipColorBackground;
+    }
+}
+
+void Mouse::Draw_Tooltip()
+{
+    if (g_theScriptEngine->Get_Fade() != ScriptEngine::FADE_NONE && m_displayTooltip && g_theDisplay != nullptr
+        && m_tooltipDisplayString != nullptr && m_tooltipDisplayString->Get_Text_Length() > 0 && !m_isTooltipEmpty) {
+        int string_x;
```
Comment (xezon) [inline src/game/client/input/mouse.cpp]: perhaps call it `string_w`, `string_h` instead.

File: src/game/client/input/mouse.cpp [hunk 8]
```diff
@@ -773,3 +773,145 @@ void Mouse::Draw_Cursor_Text() const
...
+        int string_y;
+        m_tooltipDisplayString->Get_Size(&string_x, &string_y);
+        int x = m_currMouse.pos.x + 20;
```
Comment (xezon) [inline src/game/client/input/mouse.cpp]: `mouse_x`, `mouse_y`

File: src/platform/w3dengine/client/w3dmouse.cpp (hunk 1 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dmouse.cpp]: fixed.

File: src/platform/w3dengine/client/w3dmouse.cpp (hunk 2 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dmouse.cpp]: fixed.

File: src/platform/w3dengine/client/w3dmouse.cpp (hunk 3 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dmouse.cpp]: fixed.

File: src/platform/w3dengine/client/w3dmouse.cpp (hunk 4 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dmouse.cpp]: fixed.

File: src/game/client/input/mouse.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/client/input/mouse.cpp:788]: No, this is correct.

File: src/game/client/input/mouse.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/client/input/mouse.cpp:820]: This is what the original code is doing.

File: src/game/client/input/mouse.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/client/input/mouse.cpp]: fixed.

File: src/game/client/input/mouse.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/client/input/mouse.cpp]: fixed.

File: src/platform/w3dengine/client/w3dmouse.cpp [hunk 9]
```diff
@@ -97,19 +101,133 @@ void W3DMouse::Init()
...
+        tm[1].Z = 0.0f;
+        tm[2].X = 0.0f;
+        tm[2].Y = 0.0f;
+        tm[2].Z = 1.0f;
+        tm[2].W = -1.0f;
+
+        if (g_theInGameUI != nullptr && g_theInGameUI->Is_Scrolling()) {
+            Coord2D scroll = g_theInGameUI->Get_Scroll_Amount();
+            scroll.Normalize();
+            tm.Rotate_Z(GameMath::Atan2(-scroll.y, scroll.x) - 1.5707964f);
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dmouse.cpp]: PI / 2

**Review by xezon — COMMENTED** [review]
Tool tip text is not visible.

Even with `Hook_Any(0x004035C0, Mouse::Set_Cursor_Tooltip);` commented out.

With tool tip I mean the box when hovering units, or UI elements in the Menu.

shot_20230926_164120_3

Here I hover the Map Start Position with Thyme build, and it will not show the tool tip text for it.

File: src/platform/w3dengine/client/win32mouse.h [hunk 10]
```diff
@@ -35,10 +35,12 @@ class Win32Mouse : public Mouse
     virtual ~Win32Mouse() { g_theWin32Mouse = nullptr; };
 
     virtual void Init() override;
+    virtual void Reset() override { Mouse::Reset(); }
```
Comment (duncanspumpkin) [inline src/platform/w3dengine/client/win32mouse.h:38]: ? seems a bit odd to implement a function that is defined by default. Same goes for Update.

File: src/platform/w3dengine/client/w3dmouse.cpp (hunk 9 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dmouse.cpp]: fixed.

**Review by xezon — APPROVED** [review]
Cursor and tool tip looks normal.

**Pull request merged**